File: monty_hall_clean.py
"""Python equivalent of monty_hall_clean.R."""

from __future__ import annotations

import random
import statistics
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt


OUT_DIR = Path("outputs")


def simulate(
    n_experiments: int = 1000,
    n_chances: int = 1000,
    num_doors: int = 4,
) -> tuple[list[int], list[int], dict[str, int]]:
    door_list = list(range(1, num_doors + 1))
    diff_door = 0
    switch_correct = 0
    stay_correct = 0

    success_switch_counts: list[int] = []
    success_stay_counts: list[int] = []

    for _ in range(n_experiments):
        success_switch = 0
        success_stay = 0

        for _ in range(n_chances):
            car = random.choice(door_list)
            guess = random.choice(door_list)

            possible_hall = [door for door in door_list if door not in (guess, car)]
            hall = possible_hall[0] if len(possible_hall) == 1 else random.choice(possible_hall)

            switch_options = [door for door in door_list if door not in (guess, hall)]
            switched_guess = switch_options[0] if len(switch_options) == 1 else random.choice(switch_options)

            if switched_guess == car:
                success_switch += 1
                s_switch = 1
            else:
                s_switch = 0

            if guess == car:
                success_stay += 1
                s_stay = 1
            else:
                s_stay = 0

            if s_stay == 0 and s_switch == 0:
                diff_door += 1
            if s_stay == 0 and s_switch == 1:
                switch_correct += 1
            if s_stay == 1 and s_switch == 0:
                stay_correct += 1

        success_switch_counts.append(success_switch)
        success_stay_counts.append(success_stay)

    counts = {
        "diff_door": diff_door,
        "switch_correct": switch_correct,
        "stay_correct": stay_correct,
    }
    return success_switch_counts, success_stay_counts, counts


def plot(success_switch: list[int], success_stay: list[int], n_chances: int) -> None:
    OUT_DIR.mkdir(exist_ok=True)
    fig, ax = plt.subplots(figsize=(8, 4.5))
    ax.hist(success_switch, bins=30, color="red", alpha=0.25, edgecolor="red", label="switch")
    ax.hist(success_stay, bins=30, color="blue", alpha=0.25, edgecolor="blue", label="stay")
    ax.set_title("Monty Hall: switch vs stay")
    ax.set_xlabel(f"successes out of {n_chances} per experiment")
    ax.set_ylabel("frequency")
    ax.legend()
    fig.tight_layout()
    fig.savefig(OUT_DIR / "clean_switch_vs_stay.png", dpi=160)
    plt.close(fig)


def main() -> None:
    n_experiments = 1000
    n_chances = 1000
    num_doors = 4
    n_total = n_experiments * n_chances

    success_switch, success_stay, counts = simulate(n_experiments, n_chances, num_doors)
    plot(success_switch, success_stay, n_chances)

    print(
        "Mean and SD successes with switch =",
        statistics.fmean(success_switch),
        statistics.stdev(success_switch),
    )
    print(
        "Mean and SD successes without switch =",
        statistics.fmean(success_stay),
        statistics.stdev(success_stay),
    )
    print("Entirely missed =", counts["diff_door"])
    print("Switch correct =", counts["switch_correct"])
    print("Stay correct =", counts["stay_correct"])
    print(counts["diff_door"] + counts["switch_correct"] + counts["stay_correct"])
    print("confirming N_tot =", n_total, "=", n_experiments * n_chances)


if __name__ == "__main__":
    main()

File: monty_hall_clean.R
N_experiments <- 1000
N_chances <- 1000
num_doors <- 4
door_list <- 1:num_doors

N_tot <- N_experiments * N_chances

stay_correct <- 0
switch_correct <- 0
diff_door <- 0

succ_df <- data.frame(
  n = 1:N_experiments,
  success_switch = integer(N_experiments),
  success_stay = integer(N_experiments)
)

for (e in 1:N_experiments) {
  success_switch <- 0
  success_stay <- 0
  
  for (n in 1:N_chances) {
    car <- sample(door_list, 1)
    guess <- sample(door_list, 1)
    
    # Host opens a goat door that is not the guessed door
    possible_hall <- setdiff(door_list, c(guess, car))
    
    #if (length(possible_hall) == 0) {
    #  # guess == car, so host can open either goat door
    #  possible_hall <- setdiff(door_list, guess)
    #}
    
    #hall <- sample(possible_hall, 1)
    hall <- if (length(possible_hall) == 1) possible_hall else sample(possible_hall, 1)
    
    switched_guess <- setdiff(door_list, c(guess, hall))
    
    if (length(switched_guess) == 1) {
      if (switched_guess == car)  {
        success_switch <- success_switch + 1
        s_switch <- 1
      } else {
        s_switch <- 0
      }
    } else {
      switched_guess <- sample(switched_guess, 1)
      if (switched_guess == car) {
        success_switch <- success_switch + 1
        s_switch <- 1
      } else {
        s_switch <- 0
      }
    }
    
    if (guess == car) {
      success_stay <- success_stay + 1
      s_stay <- 1
    } else {
      s_stay <- 0
    }
    
    # some elementary counting here #
    # Quantify our data 
    
    # if s_stay = 0 and s_switch = 0 <- car was in a different slot entirely. missed
    # if s_stay = 0 and s_switch = 1 <- switching was correct
    # if s_stay = 1 and s_switch = 0 <- original guess was correct
    
    if (s_stay == 0 & s_switch == 0) {
      diff_door <- diff_door + 1
    }
    
    if (s_stay == 0 & s_switch == 1) {
      switch_correct <- switch_correct + 1
    }
    
    if (s_stay == 1 & s_switch == 0) {
      stay_correct <- stay_correct + 1
    }
  } # chances in a given experiment loop here #
  
  succ_df$success_switch[e] <- success_switch
  succ_df$success_stay[e] <- success_stay
} # experiment loop here 

# plotting 
library(ggplot2)

ggplot(succ_df) +
  geom_histogram(aes(x = success_switch), bins = 30,
                 colour = "red", fill = "red", alpha = 0.25) +
  geom_histogram(aes(x = success_stay), bins = 30,
                 colour = "blue", fill = "blue", alpha = 0.25) +
  ggtitle("Monty Hall: switch vs stay") +
  xlab("Successes out of 100 per experiment")

cat("Mean and SD successes with switch = ",
    mean(succ_df$success_switch),
    sd(succ_df$success_switch), "\n")

cat("Mean and SD successes without switch = ",
    mean(succ_df$success_stay),
    sd(succ_df$success_stay), "\n")


cat("Entirely missed = ", diff_door, "\n")

cat("Switch correct = ", switch_correct, "\n")

cat("Stay correct = ", stay_correct, "\n")

diff_door + switch_correct + stay_correct
cat("confirming N_tot = ", N_tot, " = ", N_experiments*N_chances,"\n")


